dataset/src/main/scala/frameless/TypedColumn.scala

```diff
@@ -10,26 +10,81 @@ sealed trait UntypedExpression[T] {
   def expr: Expression
 }
 
```

Comment:
@adelbertc @OlivierBlanvillain @kanterov 
Below is a first experimental implementation of the TypeCoercion type class.  The idea is to allow operations with columns that have compatible numerical types (e.g., 1 + 2L, or 5L + 0.2). 

```scala
scala> val e = TypedDataset.create[(Int, String, Long)]( (1,"a",2L) :: (2, "b", 4L) :: (2, "b", 1L) :: Nil )
e: frameless.TypedDataset[(Int, String, Long)] = [_1: int, _2: string ... 1 more field]

// This is what we had so far (column and right hand expression needed to have the same type)
scala> e.select(e('_1) + 4)
res2: frameless.TypedDataset[Int] = [(_1 + 4): int]

// Here we allow the type to widen 
scala> e.select(e('_1) + 4L)
res3: frameless.TypedDataset[frameless.TypeCoercion.intLong.ResultType] = [(_1 + 4): bigint]

scala> e.select(e('_1) + 0.1)
res4: frameless.TypedDataset[frameless.TypeCoercion.intDouble.ResultType] = [(_1 + 0.1): double]

// Note how _1 and _3 had different types (Int and Long respectively)
scala> e.select(e('_1) + e('_3))
res5: frameless.TypedDataset[Long] = [(_1 + _3): bigint]

scala> e.select(e('_1) + e('_3)).collect().run()
res6: Seq[Long] = WrappedArray(3, 6, 3)
```